presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,192 @@
+                new BigIntegerAndLongStateFactory(),
+                type,
+                false);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, type, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        initializeIfNeeded(state);
+        BigInteger value = BigInteger.valueOf(type.getLong(block, position));
+        state.setBigInteger(state.getBigInteger().add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    public static void inputLongDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        initializeIfNeeded(state);
+        BigInteger value = unscaledValueToBigInteger(type.getSlice(block, position));
+        BigInteger accumulated = state.getBigInteger();
+        state.setBigInteger(accumulated.add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    {
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(BigInteger.valueOf(0));
+        }
+    }
+
+    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    {
+        state.setLong(state.getLong() + otherState.getLong());
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(otherState.getBigInteger());
+        }
+        else {
+            state.setBigInteger(state.getBigInteger().add(otherState.getBigInteger()));
+        }
+    }
+
+    public static void outputShortDecimal(ShortDecimalType type, BigIntegerAndLongState state, BlockBuilder out)
+    {
+        if (state.getLong() == 0) {
+            out.appendNull();
+        }
+        else {
+            type.writeLong(out, average(state, type).unscaledValue().longValue());
+        }
+    }
+
+    public static void outputLongDecimal(LongDecimalType type, BigIntegerAndLongState state, BlockBuilder out)
+    {
+        if (state.getLong() == 0) {
+            out.appendNull();
+        }
+        else {
+            type.writeBigDecimal(out, average(state, type));
+        }
+    }
+
+    private static BigDecimal average(BigIntegerAndLongState state, DecimalType type)
+    {
+        BigDecimal sum = new BigDecimal(state.getBigInteger(), type.getScale());
+        BigDecimal count = BigDecimal.valueOf(state.getLong());
+        return sum.divide(count, type.getScale(), ROUND_DOWN);
```

Comment:
need to check what rounding implementations in other DBs use, the SQL standard doesn't specify this, it only says that the scale and precision shouldn't be less than the input